Monads.NET/Maybe.cs

```diff
@@ -98,10 +99,44 @@ public static TInput IfDo<TInput>(this TInput o, Func<TInput, bool> evaluator, A
         public static TResult With<TKey, TResult>(this IDictionary<TKey, TResult> o, TKey key)
             where TResult : class
         {
-            return (o == null || !o.ContainsKey(key)) ? null : o[key];
+            TResult value;
+            return (o == null || !o.TryGetValue(key, out value)) ? null : value;
         }
 
         /// <summary>
+        /// Allows for a null safe accessing of an item
+        /// </summary>
+        /// <typeparam name="TInput">Reference type of object being extended</typeparam>
+        /// <typeparam name="TResult">Reference type of object being returned</typeparam>
+        /// <param name="o">instance of object being extended</param>
+        /// <param name="evaluator">function that acts on object to return a result</param>
+        /// <returns>null if object is null or an instance of TResult</returns>
+        public static Task<TResult> WithAsync<TInput, TResult>(this TInput o, Func<TInput, Task<TResult>> evaluator)
+            where TResult : class
+            where TInput : class
+        {
+            return (o == null) ? TaskEx.FromResult<TResult>(null) : evaluator(o);
+        }
+
+        /// <summary>
+        /// Allows for a null safe accessing of items
+        /// </summary>
+        /// <typeparam name="TInput">Reference type of object being extended</typeparam>
+        /// <typeparam name="TResult">Reference type of object being returned</typeparam>
+        /// <param name="o">instance of collections of objects being extended</param>
+        /// <param name="evaluator">function that acts on object to return a result</param>
+        /// <returns>null if object is null or an instance of TResult</returns>
+        public async static Task<IEnumerable<TResult>> WithAsync<TInput, TResult>(this IEnumerable<TInput> o, Func<TInput, Task<TResult>> evaluator)
+            where TResult : class
+            where TInput : class
+        {
+            if (o == null) return null;
+            
+            return await TaskEx.WhenAll(o.Select(async e => await evaluator(e)));
```

Comment:
Do we know whether the `TaskEx.WhenAll` implementation will kick all the tasks off at once, or do a blocking loop? Doing a `ToArray()` here will guarantee the tasks are started at once, and is preferred. Also same question about `TaskEx` instead of 
` await Task.Factory.ContinueWhenAll(tasks.ToArray(), i => {});`